Add ` tcp_nodelay` option and fix race condition with deadlock

kellerkindt opened the pull request:
Besides smaller cleanup commits:
 * [Fix needless_borrow clippy warning](https://github.com/lulf/dove/commit/bf430343fe525010b9a1e65b0b3e5d138b6cc038)
 * [Warn on no next_outgoing_id](https://github.com/lulf/dove/commit/6db37582228c35f0b7e7536581c9681d9c0baaac)
 * [Fix clippy lints](https://github.com/lulf/dove/commit/4844dc298b516c54ad63c0ad421adb8158e52d5b)

This PR adds a `tcp_nodelay` option to `ConnectionOptions` 
 * [Add tcp_nodelay to ConnectionOptions](https://github.com/lulf/dove/commit/23c63e3234c8e322229b9a31912ef021365114a6)
 * [Set tcp_nodelay only after the connection was established (req by windows)](https://github.com/lulf/dove/commit/0b5a5702b29f12317bcd93e722e097c8838dec01) 



And fixes a race condition that could lead to a deadlock in sending messages:
 * [Dont interfer in the removal of deliver-id mappings of sent messages](https://github.com/lulf/dove/commit/6dfe93866501e863cef7e0fd8aad8733483484d8)

I'd like to discuss the race condition fix. For me the `.remove` call on `did_to_delivery` caused by the received messages seems to be "obviously" wrong: the mapping in `did_to_delivery` is only for sent/outgoing messages. A scenerio with a lot of messages being received and transmitted seems to run into an id-clash and results in a deadlock. Consider the following scenario:
 - A message is received. [The `delivery_id` is rememberd](https://github.com/lulf/dove/blob/main/src/container.rs#L705).
 - A new message [shall be sent](https://github.com/lulf/dove/blob/main/src/container.rs#L629). The [`next_outgoing_id`](https://github.com/lulf/dove/blob/main/src/driver.rs#L680) has by chance the same value as the `delivery_id` from the just received message. [An id-mapping is added for said `next_outgoing_id`](https://github.com/lulf/dove/blob/main/src/driver.rs#L702).
 - The `Delivery`-Container of the received message [is dropped](https://github.com/lulf/dove/blob/main/src/container.rs#L769). This results in a call to [remove the id-mapping for its `delivery_id`](https://github.com/lulf/dove/blob/main/src/driver.rs#L784).
 - The send-operation now expectes an response for its tranmissions, but waits forever [here](https://github.com/lulf/dove/blob/main/src/container.rs#L637) because it cannot [be deliverd](https://github.com/lulf/dove/blob/main/src/driver.rs#L442) because of the broken mapping.

Not so sure about the removal of the `incoming_window` increment. But this seems also to be updated via `SessionFlowControl::accept` ([driver.rs:96](https://github.com/lulf/dove/blob/main/src/driver.rs#L96)), which in turn is called in [line 389](https://github.com/lulf/dove/blob/main/src/driver.rs#L389).

Could you comment on my assumptions?

kellerkindt committed:
Fix needless_borrow clippy warning

kellerkindt committed:
Add tcp_nodelay to ConnectionOptions

kellerkindt committed:
Set tcp_nodelay only after the connection was established (req by windows)

kellerkindt committed:
Warn on no next_outgoing_id

kellerkindt committed:
Dont interfer in the removal of deliver-id mappings of sent messages

This fixes an issue where incoming messages could lock up a send request
because dropping a received Delivery removes the mapping for the Sender.
The SessionDriver will then not be able to notifiy the LinkDriver in
SessionDriver::dispatch => Performative::Disposition (line 438). The
Sender then might wait forever in Sender::send_withMode =>
self.link.recv() (line 637).

On another note: This seems also seems to be an error besides the
potential lockup of the Sender. Why does a received Delivery/Message
interfere in the mappings of sent messages at all?

kellerkindt committed:
Fix clippy lints

lulf commented:
It's been a while for me since I looked at this, so could be my context is inaccurate, but I think the fix is correct. The did_to_delivery is intended to track the outgoing deliveries, and the fn disposition() is really about sending dispositions on deliveries from the other side, so it should not have any business touching the did_to_delivery. When the Disposition message is received from the remote, the did_to_delivery is removed, which is what makes sense.

As for the incoming window adjustment, I wonder if that is correct. With this change, the incoming window will only be decremented if the remote Flow sets it lower than the current value. I'm wondering if this could cause issues depending on the broker implementation, but if you have tested it beyond the max window and the state looks reasonable (i.e. not stuck at max or something).

kellerkindt committed:
Restore incoming_window incrementation in LinkDriver::disposition

This incrementation was previously removed in
6dfe93866501e863cef7e0fd8aad8733483484d8

kellerkindt commented:
Thanks for your feedback. I think your concerns about the `incoming_window` are right: it looks like it would fail after 2^31 received messages - and I didn't test for _that_ many messages :) So I undid the removal in [807fe9c](https://github.com/lulf/dove/pull/42/commits/807fe9cd599e9e4d701b89b5436b991ee48f4ec6)

lulf commented:
Thanks for testing and the fixes!
